### DISCUSSION TITLE
develop-release-candidate needs testing

### DISCUSSION THREAD

**Discussion by doudar** [doudar/SmartSpin2k#117] [General]
@kadaan and @fireyeti - Please build and test branch https://github.com/doudar/SmartSpin2k/tree/develop-release-candidate

It's a combination of @kadaan 's last commits and my rewrite of how the BLE client is handled. 

Let me know if you need binaries. I've brought a handful of esp-32's with me on this trip, but there's only so much I can do without an actual indoor riding session. 

Everything looks working to me. There's bound to be issues however. Please find them! 

Good Luck!

**Comment by fireyeti** [comment]
So I'm trying to figure out how to build the binary in Visual Studio/PlatformIO. Anyone have tips? I can see in the blue bar at the bottom of Visual Studio I am in the correct branch. What's next? Imagine you are talking to a child, as I have never built a binary. :-)

**Reply by doudar** [reply]
bottom of the screen you'll see a tiny white check mark. clicking on that will build it

**Reply by doudar** [reply]
when it's finished, there will be a firmware.bin in your computers Github/SmartSpin2k/.pio/build/esp32doit/ folder

**Reply by fireyeti** [reply]
Cool! Thanks @doudar !

**Comment by drz01** [comment]
Hmm. Is there a easy way to confirm the build came from the correct branch? I was pretty sure, but could not tell from either the console or the UI...

The resultant firmware behaved more like the last release than the last FTMS test build...did the new scanner make it into this branch?

**Reply by doudar** [reply]
You should see a much different looking firmware version at the bottom of the web page:

![image](https://user-images.githubusercontent.com/17362216/110265867-2baa7700-7f82-11eb-8bb7-0f41c7cb0ebf.jpeg)

Make sure it's not automatically updating also. I don't currently have a way to test that.

**Reply by fireyeti** [reply]
When I built the branch it displayed 1.2.15? I'll try to build again to see what happens.

**Reply by drz01** [reply]
Yes, that is the problem I am having.

**Reply by doudar** [reply]
> Yes, that is the problem I am having.

Was is auto updating? You can turn that off in the settings.

**Reply by drz01** [reply]
Finally built from branch after a manual Github sync before autosync would start; I also merged the latest commits: [Console output](https://pastebin.com/gUXEeFqu)

**Reply by doudar** [reply]
Whoa....That's ugly.

I wonder why it kept rescanning... 

Thanks for the pastebin - that works so much better for the debugging output.

**Reply by doudar** [reply]
Just for kicks, does it do the same thing using my binary uploaded either via the web interface or via the SmartSpin2K flash tool https://github.com/SmartSpin2K/SmartSpin2kFlasher/releases/

**Reply by drz01** [reply]
This is from your build: [console output](https://pastebin.com/QVY6hmv8)

I see Telegram is still not sending; inching closer though testing with my own bot.

**Reply by doudar** [reply]
I'm working from my laptop (instead of my main computer) and I don't have the telegram certificates on this one so the telegram will fail (but it shouldn't cause any additional issues). I'm only keeping telegram enabled for the debug build so it takes up the same memory footprint so I can see if everything still fits in the stack after all the changes.

**Comment by doudar** [comment]
Here's my build - you'll have to unzip it before uploading. 

[removed old binary]

**Comment by kadaan** [comment]
I think the issue is that when there is no SPIFFS filesystem data on the device it download firmware. When you setup the device use the “build” and “build filesystem image” platform lip. actions. Then you can do an “upload filesystem image” and “upload and monitor”. Watch the logs and it should skip downloading a firmware.

**Comment by doudar** [comment]
Here's a new binary (from the develop-release-candidate branch) with the latest changes from today. 

This should have fixed the scan loop.

[removed old binary]

**Reply by doudar** [reply]
Try this one:
[removed old binary]
removed WiFi.config(INADDR_NONE, INADDR_NONE, INADDR_NONE);

**Reply by fireyeti** [reply]
Loaded correctly. Will test with a ride later.

**Reply by fireyeti** [reply]
Was able to test with Zwift. Trainer and SS2K are in same location as previous. I am 15' from a WiFi Bridge AP, and the SS2K does not connect to the wifi network. I had a few issues connecting the powermeter, but I belive that was related to my iOS devices capturing the BLE signal. Once I switched off Bluetooth it was able to scan and connect. Tested on a ride and the response time 
 was very good and all functions worked properly. No new issues from previous build asidde from WiFi connection not working. I was able to connect to the main Wifi point earlier within about 10'. Will try to test out distances from the bridge AP and report back later.

**Comment by doudar** [comment]
If you want, try changing the BLE loop times to 1000 & 999 in settings.h
and see if that helps.

On Mon, Mar 8, 2021 at 2:59 PM Nick B <notifications@github.com> wrote:

> Was able to test with Zwift. Trainer and SS2K are in same location as
> previous. I am 15' from a WiFi Bridge AP, and the SS2K does not connect to
> the wifi network. I had a few issues connecting the powermeter, but I
> belive that was related to my iOS devices capturing the BLE signal. Once I
> switched off Bluetooth it was able to scan and connect. Tested on a ride
> and the response time
> was very good and all functions worked properly. No new issues from
> previous build asidde from WiFi connection not working. I was able to
> connect to the main Wifi point earlier within about 10'. Will try to test
> out distances from the bridge AP and report back later.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/doudar/SmartSpin2k/discussions/117#discussioncomment-451690>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEEO2KACHINALFIXGIQXIDDTCU3DNANCNFSM4YXPGWBQ>
> .
>
-- 

*----------------------*
*Anthony Doud*

**Comment by doudar** [comment]
Latest rebuild - Loop times increased for hopefully better WiFi performance, + sensor changes by @kadaan . 

[removed old binary]

**Reply by drz01** [reply]
FTMS not working yet on mine... [console](https://pastebin.com/Q3bwyr0j)

NB tried merging to latest commit with similar result

**Reply by kadaan** [reply]
Can you remind me what bike you have?

**Reply by drz01** [reply]
SM400 from Lifespan Fitness. Pretty generic products; I think the "brand" contracts out batches and calls them different model numbers. BLE in the bike console was undocumented! Works in Zwift though. Hmm...

**Comment by doudar** [comment]
Just curious, does the current master branch connect if you build it?

On Wed, Mar 10, 2021 at 3:35 AM Charles <notifications@github.com> wrote:

> FTMS not working yet on mine... console <https://pastebin.com/Q3bwyr0j>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/doudar/SmartSpin2k/discussions/117#discussioncomment-461115>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEEO2KEVO5SDWEGFSPZHGL3TC44N5ANCNFSM4YXPGWBQ>
> .
>
-- 

*----------------------*
*Anthony Doud*

**Reply by drz01** [reply]
From master: [console](https://pastebin.com/VxhtEpv4)

Also tried latest drc branch.

**Reply by kadaan** [reply]
@drz01 Did you already state which device from within the logs is the Bike?  If not, if you unplug the bike and do another scan, we could diff the two logs and narrow down to the one interesting device.

**Reply by drz01** [reply]
@kadaan ZV-E31E6 is the bike.

**Reply by doudar** [reply]
Latest and greatest binary. 

I tweaked some of the connection settings. It's possible (not entirely probable however) that those might enable you to connect. 

[firmware.zip](https://github.com/doudar/SmartSpin2k/files/6135676/firmware.zip)

**Reply by fireyeti** [reply]
Testing software build now.

Loads correctly, all BLE servers connect correctly and connect to Zwift.

Once BLE servers are connected and active Zwift session is running, the web pages only load intermitently. SmartSpin is connected to local network, but I cannot load the web interface consistently.

All other services seem to be running correctly.

**Reply by fireyeti** [reply]
After about 5 minutes the web interface does not load at all. SmartSpin remains connected to the local internet.

Edit to add. Reboot the SS2k, tried to connect via smartspin2k.local and no luck. I then tried the local IP and was able to connect to the web interface.

**Reply by fireyeti** [reply]
20 minutes later the web server is loading via smartspin2k.local unfortunately don't have any idea what changed.

**Reply by doudar** [reply]
I just rode it and might have experienced the same issue in the first couple minutes but I also had 3 different browsers all pulling the debugging page (Stress test)

My biggest issue is the scan/reconnect isn't quite right. I think I need to have it disconnect first to reset all connections when you press that button as sometimes the sensors are getting stuck in an intermediate state (connected but not reading).

**Reply by kadaan** [reply]
Also, be aware, we have had mDNS issue in the past.  The SS2K stopped advertising. I added some code to fix it, but it might not be 100%

**Reply by doudar** [reply]
@fireyeti , I just scanned through your telegram logs and it looks like you had a reboot about ten minutes into the ride. Did you reboot, or is it a possible crash?

**Reply by kadaan** [reply]
Odd. I’ve done a 1/2 dozen rides with no crashes. Wonder what could still be causing them, if that is what this is.

**Reply by doudar** [reply]
> Odd. I’ve done a 1/2 dozen rides with no crashes. Wonder what could still be causing them, if that is what this is. 

Try my new binary if you get time. One of the things I did was switch both BLE tasks to core 0 because it's supposed to be better if everything BLE is running on the same core (for pointers I assume). That could be causing issues with the html server as I can't remember what core that runs on (it's hopefully the opposite of the BLE.)

**Reply by fireyeti** [reply]
The 10 minute point was a reboot. I had no crashes, just no HTML on the smartspin.local which would be in line with @kadaan referencing DNS issues. Especially since I could reach it via its IP directly?

**Reply by doudar** [reply]
To clarify, the page was at least responsive if you went directly to http://192.168.x.x but if you went to http://SmartSpin2k.local , nothing, correct?

Edit - cool, I got it. I see you edit above

**Reply by drz01** [reply]
not quite yet...[console](https://pastebin.com/ARAM4bNQ)

**Reply by drz01** [reply]
"Skipping non-selected PM"?? [drc build](https://pastebin.com/pqxq8sxR)

**Reply by doudar** [reply]
It appears you have "none" selected as your power meter in the Bluetooth scanner page. Select "any" and save the drop downs. 

Skipping non-selected PM |ZV-E31E6|none

**Reply by kadaan** [reply]
> Try my new binary if you get time. One of the things I did was switch both BLE tasks to core 0 because it's supposed to be better if everything BLE is running on the same core (for pointers I assume). That could be causing issues with the html server as I can't remember what core that runs on (it's hopefully the opposite of the BLE.)

One thing to note, the stepper motor is on core 0 as well.

**Comment by fireyeti** [comment]
Yes, that's exactly right

On Sat, Mar 13, 2021 at 21:58 Anthony Doud ***@***.***> wrote:

> To clarify, the page was at least responsive if you went directly to
> http://192.168.x.x but if you went to http://SmartSpin2k.local , nothing,
> correct?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/doudar/SmartSpin2k/discussions/117#discussioncomment-479547>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASEOOUKNIOTRL3ONNJILA4TTDQJ5RANCNFSM4YXPGWBQ>
> .
>
--

**Reply by kadaan** [reply]
Crazy. I added specific code to keep pushing mDNS data so this wouldn’t happen. Would prefer to get regular DNS registration working. I had that working locally, but when it got pushed into the branch others seemed to have issues.

**Reply by doudar** [reply]
latest nightly:

I'm feeling a bit better about the scanning and reconnect handling on this one.

[firmware.zip](https://github.com/doudar/SmartSpin2k/files/6138297/firmware.zip)

**Reply by drz01** [reply]
[nightly test](https://pastebin.com/GVzWw4zS)

Thank you so much for your Herculean efforts so far! Hopefully FTMS could work for most, but I have decided a power meter is unavoidable; so will switch over to FAU in a few days -- and finally enjoy SS2K!
